File: tests/gate/votekick_test.py
#!/usr/bin/env python3
"""Vote-kick gate: four players from four addresses; the start rules
(permission, too few, immune, self), the vote through the chat and the
panel, the vote.state pushes everyone gets, the target kicked when the
threshold is reached, the cooldown, a failed vote, the cancel.

One server (31230, WD_TEST_HOOKS=1, votekick.cooldown_sec lowered so the
scenario does not wait five minutes).
"""
import sys
import time

from harness import Client, Inbox, Server, TEST_HOOKS_ENV, check, err_code, info, probe, require_server, result

PORT = 31230
IPS = ["127.0.0.1", "127.0.0.2", "127.0.0.3", "127.0.0.4"]


def vote_events(inbox, since, settle=0.5):
    return [e for e in inbox.events("vote.state", since, settle=settle)]


def main():
    require_server()
    info("=== vote-kick ===")
    config = {"votekick": {"cooldown_sec": 3, "window_sec": 15}, "limits": {"commands_per_10s": 100}}
    with Server(PORT, env=TEST_HOOKS_ENV, name="votekick", config=config) as srv:
        info("--- too few players ---")
        p1 = Client.join(PORT, name="P1", source=IPS[0])
        p2 = Client.join(PORT, name="P2", source=IPS[1])
        i1, i2 = Inbox(p1), Inbox(p2)
        probe(p1, "group.set", {"group": "trusted"})
        m = i1.chat("/votekick P2")
        check(i1.chat_lines(m) == ["A vote needs at least 4 players online."], "min_players")
        p3 = Client.join(PORT, name="P3", source=IPS[2])
        p4 = Client.join(PORT, name="P4", source=IPS[3])
        i3, i4 = Inbox(p3), Inbox(p4)
        for i in (i1, i2, i3, i4):
            i.hello()

        info("--- start rules ---")
        m = i2.chat("/votekick P3")
        check(i2.chat_lines(m) == ["You may not do that (votekick.start)."], "default has no votekick.start")
        probe(p4, "group.set", {"group": "mod"})
        m = i1.chat("/votekick P4")
        check(i1.chat_lines(m) == ["That player cannot be vote-kicked."], "a mod (level 50) is immune")
        m = i1.chat("/votekick P1")
        check(i1.chat_lines(m) == ["You cannot vote-kick yourself."], "not yourself")

        info("--- a vote passes: 4 players, 3 eligible, ceil(0.6*3) = 2 yes ---")
        marks = [i.mark() for i in (i1, i2, i3, i4)]
        m = i1.chat("/votekick P3 ramming")
        lines = i1.chat_lines(m)
        check(any(ln.startswith("P1 started a vote to kick P3 (ramming). /vote yes or /vote no, 15 s, 2 yes needed")
                  for ln in lines), "the announcement", lines)
        for i, mk in zip((i2, i3, i4), marks[1:]):
            ev = vote_events(i, mk)
            check(ev and ev[0]["event"] == "started" and ev[0]["vote"]["target"]["name"] == "P3"
                  and ev[0]["vote"]["yes"] == 1 and ev[0]["vote"]["needed"] == 2,
                  "%s got the vote.state started push" % i.client.name, ev)
        m = i3.chat("/vote yes")
        check(i3.chat_lines(m) == ["The target does not vote."], "the target cannot vote")
        m = i2.chat("/vote nope")
        check(i2.chat_lines(m) == ["Usage: /vote yes|no|cancel"], "usage")
        r = i4.request("vote.state")
        check(r.get("ok") is True and r["data"]["vote"]["yes"] == 1 and r["data"]["vote"]["eligible"] == 3,
              "vote.state from the panel", r)
        mk = i2.mark()
        r = i4.request("vote.cast", {"yes": True})
        check(r.get("ok") is True and r["data"]["vote"]["yes"] == 2 and r["data"]["running"] is False,
              "vote.cast from the panel counts and decided the vote", r)
        check(p3.kicked(5) is not None, "P3 was kicked when the threshold was reached")
        ev = vote_events(i2, mk)
        check(any(e["event"] == "passed" for e in ev), "everyone got the passed push", [e["event"] for e in ev])
        lines = i2.chat_lines(mk)
        check(any("Vote passed: P3 was kicked (2 yes)." == ln for ln in lines), "the result in the chat", lines)
        p3.close()
        rows = probe(p1, "audit.tail", {"n": 5}).get("data", {}).get("rows", [])
        check(any(r_["op"] == "votekick_start" and r_["result"] == "ok" for r_ in rows), "the start is audited", rows)

        info("--- cooldown, then a vote that fails and one that is cancelled ---")
        p3 = Client.join(PORT, name="P3", source=IPS[2])
        i3 = Inbox(p3)
        m = i1.chat("/votekick P2")
        lines = i1.chat_lines(m)
        check(lines and lines[0].startswith("Wait ") and lines[0].endswith(" s before another vote."), "cooldown", lines)
        time.sleep(3.5)
        m = i1.chat("/votekick P2 test")
        check(any("started a vote" in ln for ln in i1.chat_lines(m)), "after the cooldown a new vote starts")
        m3 = i3.chat("/vote no")
        check(i3.chat_lines(m3) == ["Vote counted: 1/2 yes."], "a no vote")
        m4 = i4.mark()
        p4.chat("/vote no")
        lines = i4.chat_lines(m4)
        check(any("Vote to kick P2 failed (1/2)." == ln for ln in lines), "no can no longer be beaten: failed", lines)
        time.sleep(3.5)
        m = i1.chat("/votekick P2 again")
        i1.chat_lines(m)
        m = i2.chat("/vote cancel")
        check(i2.chat_lines(m) == ["You may not do that (votekick.cancel)."], "cancel is the mod's")
        mk = i1.mark()
        r = i4.request("vote.cancel")
        check(r.get("ok") is True, "vote.cancel from the panel", r)
        lines = i1.chat_lines(mk)
        check(any("The vote to kick P2 was cancelled." == ln for ln in lines), "cancelled in the chat", lines)
        r = i2.request("vote.cast", {"yes": True})
        check(err_code(r) == "vote.none", "no vote to cast in", r)

        info("--- the window closes ---")
        time.sleep(3.5)
        m = i1.chat("/votekick P2 slow")
        i1.chat_lines(m)
        mk = i2.mark()
        time.sleep(16)
        lines = i2.chat_lines(mk, timeout=5)
        check(any("Vote to kick P2 failed (1/2)." == ln for ln in lines), "the window closed: failed", lines)
        check(srv.lua_errors() == [], "no Lua errors", srv.lua_errors())
        for c in (p1, p2, p3, p4):
            c.close()
    return result()


if __name__ == "__main__":
    sys.exit(main())

File: tests/unit/votekick_test.lua
-- vote-kick: start rules, the count, the outcome, cooldowns, the pushes.

local boot = require("boot")

local function setup(overrides)
    -- off by default (the spec); these tests turn it on unless told otherwise
    overrides = overrides or {}
    overrides.votekick = overrides.votekick or {}
    if overrides.votekick.enabled == nil then overrides.votekick.enabled = true end
    local W = boot(overrides)
    local players = {}
    for i = 1, 6 do players[i] = node._join(i, { name = "P" .. i, accountId = i }) end
    W.perms.set_group(players[1], "trusted")   -- may start
    W.perms.set_group(players[6], "mod")       -- immune (level 50), may cancel
    for _, p in ipairs(players) do node._emit("playerJoined", p) end
    node._told, node._sent = {}, {}
    return W, players
end

local function cmd(player, line)
    node.bus.emit("chat:command", { pid = player.id, name = line:match("^/(%S+)"), args = {}, raw = line })
end

local function vote_events()
    local out = {}
    for _, e in ipairs(node._sent_of("wd:event", "all")) do
        if e.ev == "vote.state" then out[#out + 1] = e.data end
    end
    return out
end

local tests = {}

tests.start_rules = function()
    local W, P = setup()
    cmd(P[2], "/votekick P3")
    t.match(node._told_text(2), "You may not do that %(votekick%.start%)")
    cmd(P[1], "/votekick P1")
    t.match(node._told_text(1), "cannot vote%-kick yourself")
    cmd(P[1], "/votekick P6")
    t.match(node._told_text(1), "cannot be vote%-kicked")
    cmd(P[1], "/votekick Nobody")
    t.match(node._told_text(1), "No player matches")
    t.falsy(W.votekick.running())
    -- too few players
    for i = 3, 6 do node._leave(i) end
    node._told = {}
    cmd(P[1], "/votekick P2")
    t.match(node._told_text(1), "at least 4 players")
    local W2 = setup({ votekick = { enabled = false } })
    node._told = {}
    cmd(P[1], "/votekick P2")
    t.match(node._told_text(1), "Vote%-kick is off")
    t.falsy(W2.votekick.running())
end

tests.a_vote_passes = function()
    local W, P = setup()
    cmd(P[1], "/votekick P3 ramming")
    t.truthy(W.votekick.running())
    local s = W.votekick.state()
    t.eq(s.eligible, 5)
    t.eq(s.needed, 3, "ceil(0.6 * 5)")
    t.eq(s.yes, 1, "the starter's own yes")
    t.eq(s.seconds_left, 60)
    -- everyone was told, in the chat and as a wd:event
    t.eq(#node._told, 6)
    t.match(node._told_text(2), "P1 started a vote to kick P3 %(ramming%)%. /vote yes or /vote no, 60 s, 3 yes needed")
    local ev = vote_events()
    t.eq(#ev, 1)
    t.eq(ev[1].event, "started")
    t.eq(ev[1].vote.target.name, "P3")
    -- the target cannot vote; others do
    node._told = {}
    cmd(P[3], "/vote yes")
    t.match(node._told_text(3), "The target does not vote")
    cmd(P[2], "/vote no")
    cmd(P[4], "/vote yes")
    t.match(node._told_text(4), "Vote counted: 2/3 yes")
    t.eq(#node._kicked, 0)
    cmd(P[5], "/vote yes")
    t.eq(#node._kicked, 1)
    t.eq(node._kicked[1].name, "P3")
    t.eq(node._kicked[1].reason, "vote-kicked: ramming")
    t.falsy(W.votekick.running())
    t.match(node._told_text(1), "Vote passed: P3 was kicked %(3 yes%)")
    ev = vote_events()
    t.eq(ev[#ev].event, "passed")
    -- cooldown for the starter and (were they back) the target
    node._told = {}
    cmd(P[1], "/votekick P2")
    t.match(node._told_text(1), "Wait 300 s before another vote")
    node._advance(301 * 1000)
    node._told = {}
    cmd(P[1], "/votekick P2")
    t.truthy(W.votekick.running())
    -- the audit has the start
    local rows = W.audit.tail(5)
    local starts = 0
    for _, r in ipairs(rows) do
        if r.op == "votekick_start" and r.result == "ok" then starts = starts + 1 end
    end
    t.eq(starts, 2)
end

tests.a_vote_fails_on_no_or_timeout = function()
    local W, P = setup()
    cmd(P[1], "/votekick P3")
    cmd(P[2], "/vote no")
    cmd(P[4], "/vote no")
    cmd(P[5], "/vote no")
    t.falsy(W.votekick.running(), "no cannot be beaten any more")
    t.match(node._told_text(1), "Vote to kick P3 failed %(1/3%)")
    t.eq(#node._kicked, 0)
    node._advance(301 * 1000)
    node._told = {}
    cmd(P[1], "/votekick P3")
    node._advance(59 * 1000)
    t.truthy(W.votekick.running())
    node._advance(2 * 1000)
    t.falsy(W.votekick.running(), "the window closed")
    t.match(node._told_text(2), "failed")
    -- a vote against a target who leaves ends
    node._advance(301 * 1000)
    cmd(P[1], "/votekick P3")
    node._emit("playerLeft", P[3])
    node._leave(3)
    t.falsy(W.votekick.running())
    -- one at a time
    node._advance(301 * 1000)
    cmd(P[1], "/votekick P2")
    node._told = {}
    cmd(P[1], "/votekick P4")
    t.match(node._told_text(1), "A vote is already running")
end

tests.cancel_and_the_leaving_voter = function()
    local W, P = setup()
    cmd(P[1], "/votekick P3")
    cmd(P[2], "/vote yes")
    node._told = {}
    cmd(P[4], "/vote cancel")
    t.match(node._told_text(4), "You may not do that %(votekick%.cancel%)")
    cmd(P[6], "/vote cancel")
    t.falsy(W.votekick.running())
    t.match(node._told_text(6), "Vote cancelled")
    t.match(node._told_text(1), "The vote to kick P3 was cancelled")
    node._told = {}
    cmd(P[2], "/vote yes")
    t.match(node._told_text(2), "There is no vote running")
    -- a voter who leaves takes the vote along (the count adjusts)
    node._advance(301 * 1000)
    cmd(P[1], "/votekick P3")
    cmd(P[2], "/vote yes")
    t.eq(W.votekick.state().yes, 2)
    node._emit("playerLeft", P[2])
    node._leave(2)
    t.eq(W.votekick.state().yes, 1)
    t.eq(W.votekick.state().eligible, 4)
    t.eq(W.votekick.state().needed, 3)
    -- through the panel protocol as well
    node._sent = {}
    W.protocol.handle(P[4], '{"id":1,"op":"vote.cast","data":{"yes":true}}')
    local replies = node._sent_of("wd:reply", 4)
    t.eq(replies[1].ok, true)
    t.eq(replies[1].data.vote.yes, 2)
    W.protocol.handle(P[5], '{"id":2,"op":"vote.state"}')
    t.eq(node._sent_of("wd:reply", 5)[1].data.vote.needed, 3)
end

return tests
